File: src/main/java/com/zero_delusions/dev_assignment/network/PlayerDataSynchronizer.java
package com.zero_delusions.dev_assignment.network;

import com.zero_delusions.dev_assignment.core.database.table.UserData;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPubSub;

import java.util.List;
import java.util.UUID;

class NetworkRequest {
    private final String targetServer;
    private final String payload;

    public NetworkRequest(String targetServer, String payload) {
        this.targetServer = targetServer;
        this.payload = payload;
    }

    public String sendAndWait(int timeout) {
        Jedis jedis = new JedisPool().getResource();
        // Unique response channel using UUID
        String responseChannel = "response:" + UUID.randomUUID();

        // Publish request to target server's channel with response channel
        jedis.publish(
                "server:" + targetServer + ":requests",
                payload + "|" + responseChannel
        );

        // Block and wait for response with timeout
        List<String> response = jedis.blpop(timeout, responseChannel);
        jedis.close();

        // Return response or timeout message
        return (response != null && response.size() >= 2)
                ? response.get(1)
                : "TIMEOUT";
    }
}

class NetworkBroadcast {
    private final String payload;

    public NetworkBroadcast(String payload) {
        this.payload = payload;
    }

    public void send() {
        Jedis jedis = new JedisPool().getResource();
        // Publish to broadcast channel for all servers
        jedis.publish("broadcast", payload);
        jedis.close();
    }
}

class NetworkEvent {
    private final String payload;

    public NetworkEvent(String payload) {
        this.payload = payload;
    }

    public void send() {
        Jedis jedis = new JedisPool().getResource();
        // Fire-and-forget event publication
        jedis.publish("events", payload);
        jedis.close();
    }
}

public final class PlayerDataSynchronizer {

    // ---- Server A (Initiator) ----
    public void transferUser(String userId, String targetServer) {
        Jedis redis = new JedisPool().getResource();
        try {
            // 1. Create a distributed lock
            redis.set("lock:" + userId, "locked", "NX", "EX", 30);

            // 2. Store serialized data
            String dataKey = "transfer:" + userId;
            redis.setex(dataKey, 60, serializeUserData(userId));

            // 3. Create and send network request
            NetworkRequest request = new NetworkRequest(targetServer, dataKey);

            // 4. Wait for acknowledgment
            String response = request.sendAndWait(20_000);

            if ("ACK".equals(response)) {
                // Success case: release lock and log event
                redis.del("lock:" + userId);
                new NetworkEvent("Transfer succeeded: " + userId).send();
            } else {
                // Failure case: broadcast cleanup and release lock
                new NetworkBroadcast("cleanup:" + dataKey).send();
                redis.del("lock:" + userId);
                new NetworkEvent("Transfer failed: " + userId).send();
            }
        } finally {
            redis.close();
        }
    }

    private String serializeUserData(String userId) {
        return "userData";
    }

    // ---- Server B (Receiver) ----
    public void setupRequestListener(String currentServerId) {
        Jedis jedis = new JedisPool().getResource();
        JedisPubSub pubSub = new JedisPubSub() {
            @Override
            public void onMessage(String channel, String message) {
                if (channel.equals("server:" + currentServerId + ":requests")) {
                    // Split message into components
                    String[] parts = message.split("\\|");
                    if (parts.length < 2) return;

                    String dataKey = parts[0];
                    String responseChannel = parts[1];

                    // Process the transfer request
                    String userData = jedis.get(dataKey);
                    deserialize(userData);
                    loadUserData(new UserData());

                    // Cleanup and acknowledge
                    jedis.del(dataKey);
                    jedis.publish(responseChannel, "ACK");
                }
            }
        };

        // Start listening in a dedicated thread
        new Thread(() -> {
            try {
                jedis.subscribe(pubSub, "server:" + currentServerId + ":requests");
            } finally {
                pubSub.unsubscribe();
                jedis.close();
            }
        }).start();
    }

    public void setupBroadcastListener() {
        Jedis jedis = new JedisPool().getResource();
        JedisPubSub pubSub = new JedisPubSub() {
            @Override
            public void onMessage(String channel, String message) {
                if (channel.equals("broadcast") && message.startsWith("cleanup:")) {
                    String dataKey = message.substring("cleanup:".length());
                    jedis.del(dataKey);
                }
            }
        };

        // Start broadcast listener thread
        new Thread(() -> {
            try {
                jedis.subscribe(pubSub, "broadcast");
            } finally {
                pubSub.unsubscribe();
                jedis.close();
            }
        }).start();
    }

    private UserData deserialize(String userData) {
        return new UserData();
    }

    private void loadUserData(UserData user) {
    }
}

File: src/main/kotlin/com/zero_delusions/dev_assignment/network/PlayerDataSynchronizer.kt
package com.zero_delusions.dev_assignment.network

import com.zero_delusions.dev_assignment.core.database.table.UserData
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPubSub
import java.util.UUID

class NetworkRequest(val targetServer: String, val payload: String) {
    fun sendAndWait(timeout: Int): String {
        val jedis = JedisPool().resource
        // Unique response channel
        val responseChannel = "response:${UUID.randomUUID()}"

        // Send request to Server B's channel
        jedis.publish(
            "server:$targetServer:requests",
            "$payload|$responseChannel"
        )

        // Block and wait for a response
        val response = jedis.blpop(timeout, responseChannel)
        jedis.close()

        return response?.component2() ?: "TIMEOUT"
    }
}

class NetworkBroadcast(val payload: String) {
    // Send a message to all servers (except self via filtering)
    fun send() {
        val jedis = JedisPool().resource
        jedis.publish("broadcast", payload) // All servers listen to "broadcast"
        jedis.close()
    }
}

class NetworkEvent(val payload: String) {
    // Fire-and-forget event (no response needed)
    fun send() {
        val jedis = JedisPool().resource
        jedis.publish("events", payload)
        jedis.close()
    }
}

/**
 * Redis player synchronization system pseudocode
 */
object PlayerDataSynchronizer {
    // ---- Server A (Initiator) ----
    fun transferUser(userId: String, targetServer: String) {
        val redis = JedisPool().resource
        // 1. Lock user data in Redis
        redis.set("lock:$userId", "locked", "NX", "EX", 30) // Lock for 30 seconds

        // 2. Serialize data and store in Redis
        val dataKey = "transfer:$userId"
        redis.setex(dataKey, 60, serializeUserData(userId))

        // 3. Send a NetworkRequest to Server B
        val request = NetworkRequest(
            targetServer = targetServer,
            payload = dataKey // Send the Redis key
        )

        // 4. Wait for acknowledgment
        when (val response = request.sendAndWait(timeout = 20_000)) {
            "ACK" -> {
                // Success: Unlock user and log event
                redis.del("lock:$userId")
                NetworkEvent("Transfer succeeded: $userId").send()
            }
            else -> {
                // Failure: Broadcast cleanup and unlock
                NetworkBroadcast("cleanup:$dataKey").send()
                redis.del("lock:$userId")
                NetworkEvent("Transfer failed: $userId").send()
            }
        }
    }

    fun serializeUserData(userId: String): String = "userData"

    // ---- Server B (Receiver) ----
    // Set up listener for incoming NetworkRequests
    fun setupRequestListener(currentServerId: String) {
        val jedis = JedisPool().resource
        val pubSub = object : JedisPubSub() {
            override fun onMessage(channel: String, message: String) {
                if (channel == "server:${currentServerId}:requests") {
                    // Split payload and response channel
                    val (dataKey, responseChannel) = message.split("|")

                    // Fetch data from Redis
                    val userData = deserialize(jedis.get(dataKey))
                    // Load data and send acknowledgment
                    loadUserData(userData)
                    jedis.del(dataKey)
                    jedis.publish(responseChannel, "ACK")
                }
            }
        }
        // Start listening in a background thread
        Thread { jedis.subscribe(pubSub, "server:${currentServerId}:requests") }.start()
    }

    // Set up listener for cleanup broadcasts
    // All servers listen to the "broadcast" channel
    fun setupBroadcastListener() {
        val jedis = JedisPool().resource
        val pubSub = object : JedisPubSub() {
            override fun onMessage(channel: String, message: String) {
                if (channel == "broadcast" && message.startsWith("cleanup:")) {
                    val dataKey = message.removePrefix("cleanup:")
                    jedis.del(dataKey) // Delete orphaned data
                }
            }
        }
        Thread { jedis.subscribe(pubSub, "broadcast") }.start()
    }

    fun deserialize(userId: String): UserData = UserData()
    fun loadUserData(user: UserData) {}
}